share/spack/setup-env.sh

```diff
@@ -60,21 +60,62 @@ _spack_shell_wrapper() {
         emulate -L sh
     fi
 
-    # accumulate flags meant for the main spack command
-    # the loop condition is unreadable, but it means:
-    #     while $1 is set (while there are arguments)
-    #       and $1 starts with '-' (and the arguments are flags)
+    # accumulate flags meant for spack itself preceding the command
     _sp_flags=""
-    while [ ! -z ${1+x} ] && [ "${1#-}" != "${1}" ]; do
-        _sp_flags="$_sp_flags $1"
-        shift
+    help_or_version_was_used=0
+    while [ -n "$1" ]; do
+        case "$1" in
+            --color|--config|--sorted-profile|--lines|--env-dir|--env)
+                # long flags taking required values
+                _sp_flags="$_sp_flags $1 $2"
+                shift 2
+                ;;
+            --help|--version)
+                # special early return flags
+                help_or_version_was_used=1
+                _sp_flags="$_sp_flags $1"
+                shift
+                ;;
+            --*)
+                # long flags taking no value
+                _sp_flags="$_sp_flags $1"
+                shift
+                ;;
+            -*)
+                # loop over all single character flags (e.g. -kle[env]), but
+                # skip the dash before the flag, so start at 1
+                for (( i=1 ; i<${#1}; i++ )); do
+                    flag="${1:$i:1}"
```

Comment:
Is this valid for a posix shell (`sh`) or is this a bashism?

Reply:
So yeah, this is a bashism, no clue how to do this in `sh` (are we trying to please alpine linux people or something?)

Reply:
We need to support bash, zsh, dash, and many other shells.

Reply:
Poor future people who want to add a new flag to spack and have to go through this madness. Wouldn't it be easier to have the shell wrapper pass an undocumented flag to spack like `spack --bash-shell-wrapper $@` and make spack generate a special exit code for shell-interacting commands which would cause the wrapper to eval the stdout from the command?

Then parsing is done in python and the shell scripts are less likely to diverge